plugin/azure/azure.go

```diff
@@ -0,0 +1,227 @@
+	for _, result := range *(recordSet.Response().Value) {
+		if result.RecordSetProperties.ARecords != nil {
+			for _, a := range *(result.RecordSetProperties.ARecords) {
+				rfc1035 := fmt.Sprintf("%s %d IN %s %s", *(result.RecordSetProperties.Fqdn), *(result.RecordSetProperties.TTL), "A", *(a.Ipv4Address))
+				r, err := dns.NewRR(rfc1035)
+				if err != nil {
+					log.Warningf("failed to process resource record set: %v", err.Error())
+				}
+				newZ.Insert(r)
+			}
+		}
+
+		if result.RecordSetProperties.AaaaRecords != nil {
+			for _, a := range *(result.RecordSetProperties.AaaaRecords) {
+				rfc1035 := fmt.Sprintf("%s %d IN %s %s", *(result.RecordSetProperties.Fqdn), *(result.RecordSetProperties.TTL), "AAAA", *(a.Ipv6Address))
+				r, err := dns.NewRR(rfc1035)
+				if err != nil {
+					log.Warningf("failed to process resource record set: %v", err.Error())
+				}
+				newZ.Insert(r)
+			}
+		}
+
+		if result.RecordSetProperties.NsRecords != nil {
+			for _, a := range *(result.RecordSetProperties.NsRecords) {
+				rfc1035 := fmt.Sprintf("%s %d IN %s %s", *(result.RecordSetProperties.Fqdn), *(result.RecordSetProperties.TTL), "NS", *(a.Nsdname))
+				r, err := dns.NewRR(rfc1035)
+				if err != nil {
+					log.Warningf("failed to process resource record set: %v", err.Error())
+				}
+				newZ.Insert(r)
+			}
+		}
+
+		if result.RecordSetProperties.SoaRecord != nil {
+			a := result.RecordSetProperties.SoaRecord
+			rfc1035 := fmt.Sprintf("%s %d IN %s %s", *(result.RecordSetProperties.Fqdn), *(result.RecordSetProperties.TTL), "SOA", fmt.Sprintf("%s %s %d %d %d %d %d", *(a.Host), *(a.Email), *(a.SerialNumber), *(a.RefreshTime), *(a.RetryTime), *(a.ExpireTime), *(a.MinimumTTL)))
+			r, err := dns.NewRR(rfc1035)
+			if err != nil {
+				log.Warningf("failed to process resource record set: %v", err.Error())
+			}
+			newZ.Insert(r)
+		}
+	}
+	return newZ
+}
+
+// ServeDNS uses the azure plugin to serve dns requests
+func (h *Azure) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r}
+	qname := state.Name()
+
+	zName := plugin.Zones(h.zoneNames).Matches(qname)
+	if zName == "" {
+		return plugin.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+	}
+
+	z, ok := h.zones[zName]
+	if !ok || z == nil {
+		return dns.RcodeServerFailure, nil
+	}
+
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Authoritative = true
+	var result file.Result
+	for _, hostedZone := range z {
+		h.zMu.RLock()
+		m.Answer, m.Ns, m.Extra, result = hostedZone.z.Lookup(ctx, state, qname)
+		h.zMu.RUnlock()
+
+		// record type exists for this name (NODATA).
+		if len(m.Answer) != 0 || result == file.NoData {
+			break
+		}
+	}
+
+	if len(m.Answer) == 0 && result != file.NoData && h.Fall.Through(qname) {
+		return plugin.NextOrFailure(h.Name(), h.Next, ctx, w, r)
+	}
+
+	switch result {
+	case file.Success:
+	case file.NoData:
+	case file.NameError:
+		m.Rcode = dns.RcodeNameError
+	case file.Delegation:
```

Comment:
are delegations supported in azure dns? Seeing only 4 record type are supported in this plugin, delegations seems like an advanced feature (although NS records are there)

Reply:
Deletations are supported - 
> Azure DNS allows you to host a DNS zone and manage the DNS records for a domain in Azure. 

https://docs.microsoft.com/en-us/azure/dns/dns-domain-delegation